src/app/shared/directives/email.directive.ts

```diff
@@ -1,9 +1,24 @@
-import { Directive } from "@angular/core";
+import { Directive, forwardRef } from "@angular/core";
+import { AbstractControl, NG_VALIDATORS, ValidationErrors, Validator, ValidatorFn } from "@angular/forms";
+
+export function emailValidator(): ValidatorFn {
+    return (control: AbstractControl): ValidationErrors | null => {
+        const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
+        if (emailRegex.test(control.value)) {
+            return null;
+        }
+        else return {invalidEmail: control.value}
+  }
+}
 
 @Directive({
     selector: '[emailValidator]',
-    providers: [/*Add your code here*/]
+    providers: [{ provide: NG_VALIDATORS, useExisting: EmailValidatorDirective, multi: true }]
 })
-export class EmailValidatorDirective {
-    // Add your code here
-}
+
+export class EmailValidatorDirective implements Validator{
+
+    validate(control: AbstractControl): ValidationErrors | null {
+        return emailValidator()(control);
```

Comment:
Hmmm, why not add the implementation here? Is there a reason why the `emailValidator` fn is declared and retured here?

Reply:
This is intentional to create a validator factory separately and call the return function in the Directives validate functions.
This way we can use the directive in a template based form (Login) + we can use separately only the validator factory for a reactive form , programatically inside a form component (Registration)(inside the formgroup , formControl validator)

Reply:
Okkkk